File: backend/src/routes/snaps.js
import { Router } from 'express';
import { query } from '../db/pool.js';
import { authenticate } from '../middleware/auth.js';

const router = Router();

// ── Snap a post (toggle) ──────────────────────────────────
router.post('/:postId', authenticate, async (req, res, next) => {
    try {
        const { postId } = req.params;

        // Check if already snapped
        const existing = await query(
            'SELECT id FROM snaps WHERE user_id = $1 AND post_id = $2',
            [req.user.id, postId]
        );

        if (existing.rows[0]) {
            // Un-snap
            await query('DELETE FROM snaps WHERE id = $1', [existing.rows[0].id]);
            await query('UPDATE posts SET snap_count = snap_count - 1 WHERE id = $1', [postId]);
            return res.json({ snapped: false });
        }

        // Snap
        await query(
            'INSERT INTO snaps (user_id, post_id) VALUES ($1, $2)',
            [req.user.id, postId]
        );
        await query('UPDATE posts SET snap_count = snap_count + 1 WHERE id = $1', [postId]);

        // Award points to the post author
        const post = await query('SELECT author_id FROM posts WHERE id = $1', [postId]);
        if (post.rows[0] && post.rows[0].author_id !== req.user.id) {
            await query(
                'INSERT INTO point_ledger (user_id, amount, reason, reference_id) VALUES ($1, 2, $2, $3)',
                [post.rows[0].author_id, 'snap_received', postId]
            );
            await query(
                'UPDATE users SET sanctuary_points = sanctuary_points + 2 WHERE id = $1',
                [post.rows[0].author_id]
            );
        }

        res.json({ snapped: true });
    } catch (err) {
        next(err);
    }
});

export default router;

File: backend/routes/snaps.py
from fastapi import APIRouter, Depends
import database
from auth_utils import get_current_user

router = APIRouter(prefix="/api/snaps", tags=["snaps"])


@router.post("/{post_id}")
async def toggle_snap(post_id: str, current_user: dict = Depends(get_current_user)):
    existing = await database.fetchrow(
        "SELECT id FROM snaps WHERE user_id = $1 AND post_id = $2",
        current_user["id"], post_id,
    )

    if existing:
        # Un-snap
        await database.execute("DELETE FROM snaps WHERE id = $1", existing["id"])
        await database.execute("UPDATE posts SET snap_count = snap_count - 1 WHERE id = $1", post_id)
        return {"snapped": False}

    # Snap
    await database.execute(
        "INSERT INTO snaps (user_id, post_id) VALUES ($1, $2)",
        current_user["id"], post_id,
    )
    await database.execute("UPDATE posts SET snap_count = snap_count + 1 WHERE id = $1", post_id)

    # Award points to the post author
    post = await database.fetchrow("SELECT author_id FROM posts WHERE id = $1", post_id)
    if post and post["author_id"] and str(post["author_id"]) != current_user["id"]:
        await database.execute(
            "INSERT INTO point_ledger (user_id, amount, reason, reference_id) VALUES ($1, 2, 'snap_received', $2)",
            post["author_id"], post_id,
        )
        await database.execute(
            "UPDATE users SET sanctuary_points = sanctuary_points + 2 WHERE id = $1",
            post["author_id"],
        )

    return {"snapped": True}
